src/openrct2-ui/windows/Staff.cpp

```diff
@@ -119,1240 +121,1169 @@ static rct_widget* window_staff_page_widgets[] = {
-static void WindowStaffOptionsMousedown(rct_window* w, rct_widgetindex widgetIndex, rct_widget* widget);
-static void WindowStaffOptionsDropdown(rct_window* w, rct_widgetindex widgetIndex, int32_t dropdownIndex);
-
-static void WindowStaffStatsMouseup(rct_window* w, rct_widgetindex widgetIndex);
-static void WindowStaffStatsResize(rct_window* w);
-static void WindowStaffStatsUpdate(rct_window* w);
-static void WindowStaffStatsInvalidate(rct_window* w);
-static void WindowStaffStatsPaint(rct_window* w, rct_drawpixelinfo* dpi);
-static void WindowStaffStatsTabPaint(rct_window* w, rct_drawpixelinfo* dpi);
-
-void WindowStaffSetColours();
-
-// 0x992AEC
-static rct_window_event_list window_staff_overview_events([](auto& events) {
-    events.close = &WindowStaffOverviewClose;
-    events.mouse_up = &WindowStaffOverviewMouseup;
-    events.resize = &WindowStaffOverviewResize;
-    events.mouse_down = &WindowStaffOverviewMousedown;
-    events.dropdown = &WindowStaffOverviewDropdown;
-    events.update = &WindowStaffOverviewUpdate;
-    events.tool_update = &WindowStaffOverviewToolUpdate;
-    events.tool_down = &WindowStaffOverviewToolDown;
-    events.tool_abort = &WindowStaffOverviewToolAbort;
-    events.text_input = &WindowStaffOverviewTextInput;
-    events.viewport_rotate = &WindowStaffOverviewViewportRotate;
-    events.invalidate = &WindowStaffOverviewInvalidate;
-    events.paint = &WindowStaffOverviewPaint;
-});
-
-// 0x992B5C
-static rct_window_event_list window_staff_options_events([](auto& events) {
-    events.mouse_up = &WindowStaffOptionsMouseup;
-    events.resize = &WindowStaffStatsResize;
-    events.mouse_down = &WindowStaffOptionsMousedown;
-    events.dropdown = &WindowStaffOptionsDropdown;
-    events.unknown_05 = &WindowStaffUnknown05;
-    events.update = &WindowStaffOptionsUpdate;
-    events.invalidate = &WindowStaffOptionsInvalidate;
-    events.paint = &WindowStaffOptionsPaint;
-});
-
-// 0x992BCC
-static rct_window_event_list window_staff_stats_events([](auto& events) {
-    events.mouse_up = &WindowStaffStatsMouseup;
-    events.resize = &WindowStaffStatsResize;
-    events.unknown_05 = &WindowStaffUnknown05;
-    events.update = &WindowStaffStatsUpdate;
-    events.invalidate = &WindowStaffStatsInvalidate;
-    events.paint = &WindowStaffStatsPaint;
-});
-
-static rct_window_event_list* window_staff_page_events[] = {
-    &window_staff_overview_events,
-    &window_staff_options_events,
-    &window_staff_stats_events,
-};
-
-static EntertainerCostume _availableCostumes[static_cast<uint8_t>(EntertainerCostume::Count)];
-
-static Staff* GetStaff(rct_window* w)
+class StaffWindow final : public Window
 {
-    auto staff = GetEntity<Staff>(EntityId::FromUnderlying(w->number));
-    if (staff == nullptr)
-    {
-        window_close(w);
-        return nullptr;
-    }
-    return staff;
-}
+private:
+    EntertainerCostume _availableCostumes[static_cast<uint8_t>(EntertainerCostume::Count)];
+    Staff* _staff;
```

Comment:
```suggestion
    Staff* _staff{};
```

Reply:
You should never have the staff pointer it can not be trusted to point to the same item between ticks. Only ever save the entity id and fetch the entity using get entity.